### PR TITLE
networkd: Add T1 and T2 DHCP options to expose in dbus API

### PR DESCRIPTION
**Pull request by naraghavan**
Include T1 and T2 DHCP options to expose in dbus API. This will be added to the JSON output when we query
org.freedesktop.network1.Manager object

### CODE REVIEW

File: src/network/networkd-json.c [hunk 1]
```diff
@@ -5,11 +5,13 @@
 #include "dhcp-server-internal.h"
 #include "dns-domain.h"
 #include "ip-protocol-list.h"
+#include "dhcp6-protocol.h"
```
Comment (bluca) [inline src/network/networkd-json.c]: keep sorted alphabetically

File: src/network/networkd-json.c [hunk 2]
```diff
@@ -47,6 +51,24 @@ static int address_build_json(Address *address, JsonVariant **ret) {
         if (r < 0)
                 return r;
 
+        /*
+         * Get T1 and T2 values only when config source is DHCP
+         */
+        if(strcmp(network_config_source_to_string(address->source), "DHCPv6") == 0) {
```
Comment (bluca) [inline src/network/networkd-json.c]: missing space between `if` and brackets, here and below

File: src/network/networkd-json.c [hunk 3]
```diff
@@ -47,6 +51,24 @@ static int address_build_json(Address *address, JsonVariant **ret) {
...
+                r = dhcp6_lease_get_lifetime(address->link->dhcp6_lease, &lifetime_t1, &lifetime_t2, &lifetime_valid);
```
Comment (bluca) [inline src/network/networkd-json.c]: lifetime_valid is not used, just drop it and pass NULL here

**Review by bluca — CHANGES_REQUESTED** [review]

File: src/network/networkd-json.c (hunk 1 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Changed in alphabetically order

File: src/network/networkd-json.c (hunk 2 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Fixed it.

File: src/network/networkd-json.c (hunk 3 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Fixed it.

File: src/network/networkd-json.c [hunk 4]
```diff
@@ -2,14 +2,16 @@
...
+#include "dhcp6-protocol.h"
+#include "dhcp6-lease-internal.h"
 #include "dhcp-server-internal.h"
 #include "dns-domain.h"
 #include "ip-protocol-list.h"
 #include "netif-util.h"
 #include "networkd-address.h"
 #include "networkd-dhcp-common.h"
 #include "networkd-json.h"
-#include "networkd-link.h"
```
Comment (bluca) [inline src/network/networkd-json.c]: why is this removed?

File: src/network/networkd-json.c [hunk 5]
```diff
@@ -47,6 +51,24 @@ static int address_build_json(Address *address, JsonVariant **ret) {
         if (r < 0)
                 return r;
 
+        /*
+         * Get T1 and T2 values only when config source is DHCP
+         */
+        if (strcmp(network_config_source_to_string(address->source), "DHCPv6") == 0) {
+                r = dhcp6_lease_get_lifetime(address->link->dhcp6_lease, &lifetime_t1, &lifetime_t2, NULL);
```
Comment (bluca) [inline src/network/networkd-json.c]: `/* ret_valid= */ NULL` instead of just `NULL`

File: src/network/networkd-json.c (hunk 4 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Sorry, this got removed accidentally. Fixed it now

File: src/network/networkd-json.c (hunk 5 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Fixed it.

**Comment by bluca** [comment]
@yuwata do we have any test coverage for the json output?

File: src/network/networkd-json.c [hunk 6]
```diff
@@ -47,6 +52,24 @@ static int address_build_json(Address *address, JsonVariant **ret) {
...
 
+        /*
+         * Get T1 and T2 values only when config source is DHCP
+         */
+        if (strcmp(network_config_source_to_string(address->source), "DHCPv6") == 0) {
+                r = dhcp6_lease_get_lifetime(address->link->dhcp6_lease, &lifetime_t1, &lifetime_t2, /*ret_valid= */ NULL);
+                if (r < 0)
+                        return r;
+        } else if (strcmp(network_config_source_to_string(address->source), "DHCPv4") == 0) {
+                r = sd_dhcp_lease_get_t1(address->link->dhcp_lease, &t1);
```
Comment (yuwata) [inline src/network/networkd-json.c]: This is not usec, but sec.

File: src/network/networkd-json.c [hunk 7]
```diff
@@ -62,6 +85,8 @@ static int address_build_json(Address *address, JsonVariant **ret) {
                                 JSON_BUILD_PAIR_FINITE_USEC("PreferredLifetimeUsec", address->lifetime_preferred_usec), /* for backward compat */
                                 JSON_BUILD_PAIR_FINITE_USEC("ValidLifetimeUSec", address->lifetime_valid_usec),
                                 JSON_BUILD_PAIR_FINITE_USEC("ValidLifetimeUsec", address->lifetime_valid_usec), /* for backward compat */
+                                JSON_BUILD_PAIR_FINITE_USEC("T1", lifetime_t1),
+                                JSON_BUILD_PAIR_FINITE_USEC("T2", lifetime_t2),
```
Comment (yuwata) [inline src/network/networkd-json.c]: T1 and T2 are not (only) for address but DHCP lease. We should add, say, `DHCPLease` field, instead.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by yuwata** [comment]
> @yuwata do we have any test coverage for the json output?

Currently, we have only coverage tests. We definitely need to test that the json output has expected values, but currently we do not have such.

File: src/network/networkd-json.c (hunk 7 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: As "Address" field already has most of the DHCP lease options, won't it be appropriate to keep T1 and T2 in the Address field.  Please let me know if you think otherwise.
Sample output from JSON
{\"Family\":10,\"Address\":[32,1,0,0,0,0,0,0,0,0,0,0,0,0,0,16],\"PrefixLength\":128,\"Scope\":0,\"ScopeString\":\"global\",\"Flags\":512,\"FlagsString\":\"no-prefixroute\",**\"PreferredLifetimeUsec\":691202932072,\"ValidLifetimeUsec\":1036802932072,\"T1\":345600000000,\"T2\":552960000000**,\"ConfigSource\":\"DHCPv6\",\"ConfigState\":\"configured\",\"ConfigProvider\":[32,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1]}]

Comment (yuwata) [reply src/network/networkd-json.c]: I mean, we should not add T1 and T2 in Address field. That's irrelevant for addresses, especially when it is assigned based on DHCP6 lease. The lifetimes are set based on IA_NA option.

Comment (naraghavan) [reply src/network/networkd-json.c]: Ok understood. So if I understand the correct way would be to have DHCP6 field which will hold T1 and T2 values.

**Comment by bluca** [comment]
> > @yuwata do we have any test coverage for the json output?
> 
> Currently, we have only coverage tests. We definitely need to test that the json output has expected values, but currently we do not have such.

Ok, we can leave it as-is for now then

File: src/network/networkd-json.c (hunk 7 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Added separate DHCPv6lease field in json which includes T1 and T2, something like below.
"DHCPv6lease\":{\"T1\":345600000000,\"T2\":552960000000}

**Review by bluca — APPROVED** [review]

File: src/network/networkd-json.c [hunk 8]
```diff
@@ -1023,6 +1025,25 @@ static int dhcp_server_append_json(Link *link, JsonVariant **v) {
...
+        if (!link->dhcp6_lease)
+                return 0;
+
+        r = json_build(&w, JSON_BUILD_OBJECT(
+                                JSON_BUILD_PAIR_FINITE_USEC("T1", link->dhcp6_lease->lifetime_t1),
+                                JSON_BUILD_PAIR_FINITE_USEC("T2", link->dhcp6_lease->lifetime_t2)));
+        if (r < 0)
+                return r;
+
+        return json_append_one(v, "DHCPv6lease", w);
```
Comment (yuwata) [inline src/network/networkd-json.c]: We have `DHCPServer` key in the above. So, I think it is better to nest one more level,
```
DHCPv6Client -> Lease -> T1, T2
```
then, we can later expose e.g. client state or so under `DHCPv6Client` field.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks much better now. But see the suggestion below.

File: src/network/networkd-json.c (hunk 8 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Updated. Now the output would look like below in JSON.

\"DHCPv6Client\":{\"Leases\":{\"T1\":345600000000,\"T2\":552960000000}}

File: src/network/networkd-json.c [hunk 9]
```diff
@@ -1023,6 +1025,43 @@ static int dhcp_server_append_json(Link *link, JsonVariant **v) {
...
+}
+
+static int dhcp6_client_append_json(Link *link, JsonVariant **v) {
+        _cleanup_(json_variant_unrefp) JsonVariant *w = NULL;
+        int r;
+
+        assert(link);
+        assert(v);
+
+        if (!link->dhcp6_lease)
```
Comment (yuwata) [inline src/network/networkd-json.c]: This should be `dhcp6_client`, instead of `dhcp6_lease`.

File: src/network/networkd-json.c [hunk 10]
```diff
@@ -1023,6 +1025,43 @@ static int dhcp_server_append_json(Link *link, JsonVariant **v) {
...
+                return 0;
+
+        r = dhcp6_client_lease_append_json(link, &w);
+
+        if (r < 0)
```
Comment (yuwata) [inline src/network/networkd-json.c]: Please drop the empty line.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-json.c (hunk 9 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Updated the code to have one more level. The output would look like this.
\"DHCPv6Client\":{\"Leases\":{\"T1\":345600000000,\"T2\":552960000000}}

Comment (naraghavan) [reply src/network/networkd-json.c]: Understood. I was in assumption that this field will be used only for lease data. I have fixed it now.

File: src/network/networkd-json.c (hunk 10 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: Fixed it.

File: src/network/networkd-json.c [hunk 11]
```diff
@@ -3,6 +3,8 @@
 #include <linux/nexthop.h>
 
 #include "dhcp-server-internal.h"
+#include "dhcp6-protocol.h"
+#include "dhcp6-lease-internal.h"
```
Comment (yuwata) [inline src/network/networkd-json.c]: `dhcp6-protocol.h` seems not necessary. At least, the two headers are not sorted.

**Review by yuwata — APPROVED** [review]
LGTM. One minor comment.

File: src/network/networkd-json.c (hunk 11 above)
Comment (naraghavan) [reply src/network/networkd-json.c]: We need dhcp6-protocol.h as we refer to "DHCP6Message"  in dhcp6-lease-internal.h else the compilation would fail.

Comment (yuwata) [reply src/network/networkd-json.c]: I see, then please include `dhcp6-protocol.h` in `dhcp6-lease-internal.h`.

Comment (naraghavan) [reply src/network/networkd-json.c]: Yes, added dhcp6-protocol.h to dhcp6-lease-internal.h

**Review by yuwata — APPROVED** [review]
LGTM.
